### PR TITLE
[EAK-504] Limited Parsys: add functionality to limit parsys-like components child component count

### PR DESCRIPTION
**Pull request by ala-n**
## Feature Final Scope:

A clientlib is provided to address the maximum children limit for containers (e.g., parsys) using a custom `cq:listener` or the design/policy property. The client library decorates the `INSERT` action and monitors editable placeholders (e.g., newpar). When the limit is reached, the `INSERT` action is hidden and inactive, placeholders are disabled, and an additional `data-max-children-marker` attribute is applied to placeholders for visual appearance customization (such as a custom red message and hover state, provided out-of-the-box).

In addition to the JS API, the EAK annotation set has been expanded with the `@MaxChildren` annotation to declare the maximum children resolver `cq:listener` using component Java metadata.

## Notes for the Current PR:

- The support for complex rules in the Java API is beyond the scope of the current issue.
- Documentation, except for JS/Java docs, is not included and will be provided via a sub-PR or separately.
- The `policy-management` and the new `max-children` functionality have been consolidated into a single `eak.policy` client library/module, as they will reuse the same logic for rule filtering.

Closes: #504

### CODE REVIEW

**Comment by ala-n** [comment]
Current TODO list:
 - Rename limited-parsys to max-children in JS API
 - Get rid of isParsys check if possible, add the ability to handle other placeholders, then 'newpar'
 - Move the isEditConfig check to a generic place from MaxChildrenHandler or remove todo
 - Investigate the possibility of displaying a message if the limit is riched

File: ui.apps/src/main/content/jcr_root/apps/etoolbox-authoring-kit/policies/js/maxChildren.js [hunk 1]
```diff
@@ -0,0 +1,130 @@
...
+ * if there are more child components inserted than the {@link MaxChildrenLimiter.LIMIT_RESOLVER_NAME} listener returns or
+ * defined by {@link MaxChildrenLimiter.LIMIT_RESOLVER_PROPERTY} in policies or designs.
+ */
+(function (ns, $, author) {
+    'use strict';
+
+    ns.MaxChildrenLimiter = ns.MaxChildrenLimiter || {};
+
+    /** The name of listener to resolve parsys children limit */
+    ns.MaxChildrenLimiter.LIMIT_RESOLVER_NAME = 'resolvemaxchildern';
```
Comment (NastaLeo) [inline ui.apps/src/main/content/jcr_root/apps/etoolbox-authoring-kit/policies/js/maxChildren.js:26]:
```suggestion
    ns.MaxChildrenLimiter.LIMIT_RESOLVER_NAME = 'resolvemaxchildren';
```

File: ui.apps/src/main/content/jcr_root/apps/etoolbox-authoring-kit/policies/js/maxChildren.js [hunk 2]
```diff
@@ -0,0 +1,130 @@
...
+     */
+    ns.MaxChildrenLimiter.getChildrenLimit = function getChildrenLimit(editable) {
+        const limit = ns.EAKUtils.executeListener(editable, ns.MaxChildrenLimiter.LIMIT_RESOLVER_NAME);
+        if (typeof limit === 'number' || typeof limit === 'string') return +limit;
+        return ns.MaxChildrenLimiter.resolveChildrenLimitFromPolicy(editable);
+    };
+
+    /**
+     * @param editable
+     * @returns {number} - children limit for the given editable
```
Comment (NastaLeo) [inline ui.apps/src/main/content/jcr_root/apps/etoolbox-authoring-kit/policies/js/maxChildren.js:55]: mb it makes sense to precise in comments what the difference from the above method

File: ui.apps/src/main/content/jcr_root/apps/etoolbox-authoring-kit/policies/js/maxChildren.js [hunk 3]
```diff
@@ -0,0 +1,130 @@
...
+    };
+
+    /**
+     * @param editable
+     * @returns {number} current children count for the given editable
+     */
+    ns.MaxChildrenLimiter.getChildrenCount = function getChildrenCount(editable) {
+        if (!editable.dom) return 0;
+        const children = editable.dom.children(':not(cq, .par, .newpar, .iparys_inherited)');
+        return children ? children.length : 0;
```
Comment (NastaLeo) [inline ui.apps/src/main/content/jcr_root/apps/etoolbox-authoring-kit/policies/js/maxChildren.js:69]:
```suggestion
        return children.length || 0;
```

File: plugin/src/main/java/com/exadel/aem/toolkit/plugin/handlers/common/MaxChildrenHandler.java [hunk 4]
```diff
@@ -0,0 +1,76 @@
...
+import com.exadel.aem.toolkit.api.annotations.policies.MaxChildren;
+import com.exadel.aem.toolkit.api.annotations.policies.PolicyTarget;
+import com.exadel.aem.toolkit.api.handlers.Handler;
+import com.exadel.aem.toolkit.api.handlers.Source;
+import com.exadel.aem.toolkit.api.handlers.Target;
+import com.exadel.aem.toolkit.plugin.utils.DialogConstants;
+
+/**
+ * Implements {@code BiConsumer} to populate a {@link Target} instance with properties originating from a {@link Source}
+ * object. The source refers to the {@code resolvemaxchildern} listener of the {@code
```
Comment (NastaLeo) [inline plugin/src/main/java/com/exadel/aem/toolkit/plugin/handlers/common/MaxChildrenHandler.java:26]:
```suggestion
 * object. The source refers to the {@code resolvemaxchildren} listener of the {@code
```

File: plugin/src/main/java/com/exadel/aem/toolkit/plugin/handlers/common/MaxChildrenHandler.java [hunk 5]
```diff
@@ -0,0 +1,76 @@
...
+ * cq:editConfig} node of an AEM component
+ */
+public class MaxChildrenHandler implements Handler {
+    /**
+     * Name of the listener to resolve max children limit for `limited-parsys` js module
+     */
+    public static final String MAX_LIMIT_RESOLVER_NAME = "resolvemaxchildern";
```
Comment (NastaLeo) [inline plugin/src/main/java/com/exadel/aem/toolkit/plugin/handlers/common/MaxChildrenHandler.java:33]:
```suggestion
    public static final String MAX_LIMIT_RESOLVER_NAME = "resolvemaxchildren";
```

**Review by NastaLeo — APPROVED** [review]

File: ui.apps/src/main/content/jcr_root/apps/etoolbox-authoring-kit/policies/js/maxChildren.js (hunk 3 above)
Comment (ala-n) [reply ui.apps/src/main/content/jcr_root/apps/etoolbox-authoring-kit/policies/js/maxChildren.js:69]: Nope, that was, actually, a check for `children = null`, but it is useless as well ) 
I will take a look, thanks!

File: ui.apps/src/main/content/jcr_root/apps/etoolbox-authoring-kit/policies/js/maxChildren.js [hunk 6]
```diff
@@ -0,0 +1,132 @@
...
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+/**
+ * Utility that disables insertion, drag/drop and copy/paste actions for container component
```
Comment (dshovchko) [inline ui.apps/src/main/content/jcr_root/apps/etoolbox-authoring-kit/policies/js/maxChildren.js:16]:
```suggestion
 * Utility that disables insertion, drag/drop, and copy/paste actions for container component
```

File: ui.apps/src/main/content/jcr_root/apps/etoolbox-authoring-kit/policies/js/maxChildren.js [hunk 7]
```diff
@@ -0,0 +1,132 @@
...
+
+    /**
+     * @param {Editable} editable
+     * @returns {boolean} true if editable is a 'newpar' parsys zone
+     */
+    ns.MaxChildrenLimiter.isPlaceholder = (editable) => editable && editable.type && editable.type.endsWith('newpar');
+
+    /**
+     * Resolves max children limit for the given editable using {@link LIMIT_RESOLVER_NAME} listener.
+     * If listener does not exist or return a number, then {@link resolveChildrenLimitFromPolicy} is used.
```
Comment (dshovchko) [inline ui.apps/src/main/content/jcr_root/apps/etoolbox-authoring-kit/policies/js/maxChildren.js:45]:
```suggestion
     * If a listener does not exist or return a number, then {@link resolveChildrenLimitFromPolicy} is used.
```

File: ui.apps/src/main/content/jcr_root/apps/etoolbox-authoring-kit/policies/js/maxChildren.js [hunk 8]
```diff
@@ -0,0 +1,132 @@
...
+     * @param editable
+     * @returns {number} current children count for the given editable
+     */
+    ns.MaxChildrenLimiter.getChildrenCount = function getChildrenCount(editable) {
+        if (!editable.dom) return 0;
+        return editable.dom.children(':not(cq, .par, .newpar, .iparys_inherited)').length;
+    };
+
+    /**
+     * Checks if editable contains equal or less children than defined in 'childrenLimit' property
```
Comment (dshovchko) [inline ui.apps/src/main/content/jcr_root/apps/etoolbox-authoring-kit/policies/js/maxChildren.js:75]:
```suggestion
     * Checks if editable contains equal or fewer children than defined in 'childrenLimit' property
```

File: ui.apps/src/main/content/jcr_root/apps/etoolbox-authoring-kit/utils/js/utils.js [hunk 9]
```diff
@@ -0,0 +1,59 @@
...
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+(function (ns) {
+    'use strict';
+
+    /**
+     * Wrap the given function with the provided wrapper. Do not wrap twice.
```
Comment (dshovchko) [inline ui.apps/src/main/content/jcr_root/apps/etoolbox-authoring-kit/utils/js/utils.js:18]:
```suggestion
     * Wraps the given function with the provided wrapper. Does not wrap twice.
```

File: ui.apps/src/main/content/jcr_root/apps/etoolbox-authoring-kit/utils/js/utils.js [hunk 10]
```diff
@@ -0,0 +1,59 @@
...
+                return false;
+            }
+        };
+        fnDecorated.original = fn;
+        fnDecorated.wrapper = wrapper;
+        return fnDecorated;
+    }
+
+    /**
+     * Execute listener for the given editable
```
Comment (dshovchko) [inline ui.apps/src/main/content/jcr_root/apps/etoolbox-authoring-kit/utils/js/utils.js:39]:
```suggestion
     * Executes listener for the given editable
```

**Review by abarmina — APPROVED** [review]

**Review by liubou-masiuk — APPROVED** [review]

File: plugin/src/main/java/com/exadel/aem/toolkit/plugin/handlers/common/MaxChildrenHandler.java [hunk 11]
```diff
@@ -0,0 +1,76 @@
...
+/**
+ * Implements {@code BiConsumer} to populate a {@link Target} instance with properties originating from a {@link Source}
+ * object. The source refers to the {@code resolvemaxchildren} listener of the {@code
+ * cq:editConfig} node of an AEM component
+ */
+public class MaxChildrenHandler implements Handler {
+    /**
+     * Name of the listener to resolve max children limit for `policies` js module
+     */
+    public static final String MAX_LIMIT_RESOLVER_NAME = "resolvemaxchildren";
```
Comment (smiakchilo) [inline plugin/src/main/java/com/exadel/aem/toolkit/plugin/handlers/common/MaxChildrenHandler.java:33]: Probably, `MAX_CHILDREN` or just `LIMIT`, not "max limit")
The same in the Javadoc comments below. The word "max" already includes the meaning "limit". If something is "max", no one can go beyond this value. Therefore, this value is by def a limit...

**Review by smiakchilo — APPROVED** [review]
🔥

**Pull request merged**